tests/e2e/autoscale/wait-until-hpa-ready.go

```diff
@@ -0,0 +1,109 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/spf13/pflag"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+)
+
+func main() {
+	var hpaName string
+	var timeout int
+	var kubeconfigPath string
+
+	defaultKubeconfigPath := filepath.Join(homedir.HomeDir(), ".kube", "config")
+
+	pflag.IntVar(&timeout, "timeout", 600, "The timeout for the check.")
+	pflag.StringVar(&hpaName, "hpa", "", "HPA to check")
+	pflag.StringVar(&kubeconfigPath, "kubeconfig-path", defaultKubeconfigPath, "Absolute path to the KubeconfigPath file")
+	pflag.Parse()
+
+	if len(hpaName) == 0 {
+		fmt.Println("hpa flag is mandatory")
+		os.Exit(1)
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+	if err != nil {
+		fmt.Printf("Error reading the kubeconfig: %s", err)
+		os.Exit(1)
+	}
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	namespace, err := client.CoreV1().Namespaces().Get(context.Background(), os.Getenv("NAMESPACE"), metav1.GetOptions{})
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	hpaClientV2 := client.AutoscalingV2().HorizontalPodAutoscalers(namespace.Name)
+	hpaClientV1 := client.AutoscalingV1().HorizontalPodAutoscalers(namespace.Name)
+
+	pollInterval := time.Second
+
+	// Search in v2 and v1 for an HPA with the given name
+	err = wait.Poll(pollInterval, 0, func() (done bool, err error) {
```

Comment:
I am still not sure why actually v1 and v2 HPAs are created in a test. 

My understanding is that only a single HPA version should be used in a given cluster. Could you please explain why both are created?

Reply:
I didn't write the test. I'm just trying to make it work on OpenShift. But this is what I found while checking for the purpose of this E2E test:
* The test creates 2 `OpenTelemetryCollector` instances to test 2 ways of creating HPAs. From the `00-install.yaml` file:
```yaml
# This creates two different deployments:
#   * The first one will be used to see if we scale properly
#   * The second is to check the targetCPUUtilization option
```
* This creates 2 HPAs (I wait for their creation and metrics reporting in steps 1 and 2)
* We start `tracegen` in step 3 and wait for one of the `OpenTelemetryCollector` instances to scale up to 2 replicas
* We remove the `tracegen` deployment to stop reporting traces in step 4
* Wait until the `OpenTelemetryCollector` scales down in step 5

When the HPAs are created, they will be created using `autoscaling/v1` or `autoscaling/v2beta2`. If you check how this test was written before my changes, you can see how in `00-assert.yaml` , the test tries to assert the `simplest-collector` HPA with `autoscaling/v1` and the `simplest-set-utilization-collector` HPA with `autoscaling/v2beta2`.

When I ran this in OpenShift 4.11, both of them were created using `autoscaling/v2beta2`  (as I pointed [in this comment](https://github.com/open-telemetry/opentelemetry-operator/issues/1364#issue-1530344787)). 

So, since in KUTTL there is no way to conditionally check for one resource or another, I created the `wait-until-hpa-ready.go` script to (given a name) dynamically:
* Look for an HPA in the `autoscaling/v2beta2` API. If found, check if the HPA status is different from unknown 
* If the HPA was not found in `autoscaling/v2beta2`, look for it in `autoscaling/v1`. If found, check if the HPA status is different from unknown

Another thing: why the HPAs are created using different `autoscaling` API versions (as we can see in `00-assert.yaml`) in the Kubernetes versions tested during the CI? I think this is because one is setting the `.spec.minReplicas` and `.spec.maxReplicas` values and the other is setting them in `.spec.autoscaler` (as the comment in `00-install.yaml` explains). If you want, I can do a deeper investigation about why this happens, but in a separate issue since is not related to the current PR.